### PR TITLE
Restore WebDAV functionality

### LINKED ISSUE #744: Re-integrate WebDAV support
Related to #649 , #696 , #742 

When ZServer was moved out of Zope the webdav support module was removed as well, even though it didn't neeed to. Bringing it back will restore WebDAV functionality to Zope.

The removal was done in 24d566b475ec60a5f872d39cdb63cb730fdc0f82, PR #696

### PR DESCRIPTION
**Pull request by dataflake**
Fixes #744 

Reverts most of #696 as well as the the webdav part of 24d566b475ec60a5f872d39cdb63cb730fdc0f82

Related to #649, #742 and to 

This PR has been put together from several sources. The first was the move of the `webdav` package into ZServer (24d566b475ec60a5f872d39cdb63cb730fdc0f82) and then the removal of all remaining support code in #696. I have left out anything that relates to FTP only, as that is an entirely separate protocol. Some inspiration was taken from https://github.com/datakurre/ZServer/tree/py3 by @datakurre. 

With the `webdav` package back in Zope supporting the refreshed ZServer code by @datakurre   should become easier as well.

### CODE REVIEW

File: src/ZPublisher/BaseRequest.py [hunk 1]
```diff
@@ -451,6 +457,19 @@ def traverse(self, path, response=None, validated_hook=None):
...
+                    # This is webdav support. The last object in the path
+                    # should not be acquired. Instead, a NullResource should
+                    # be given if it doesn't exist:
+                    if no_acquire_flag and \
+                       hasattr(object, 'aq_base') and \
+                       not hasattr(object, '__bobo_traverse__'):
+
+                        if (object.__parent__ is not
+                                aq_inner(object).__parent__):
+                            from webdav.NullResource import NullResource
```
Comment (icemac) [inline src/ZPublisher/BaseRequest.py]: Could this import be moved to the top?

File: src/webdav/PropertySheets.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+from App.Common import rfc1123_date
+from OFS.interfaces import IWriteLock
+from OFS.PropertySheets import PropertySheet
+from OFS.PropertySheets import View
+from OFS.PropertySheets import Virtual
+from webdav.common import isDavCollection
+from webdav.common import urlbase
+
+
+def absattr(attr):
```
Comment (icemac) [inline src/webdav/PropertySheets.py]: There is another identical implementation in `webdav/common.py`.

**Comment by icemac** [comment]
Should we even backport this PR to Zope 4 to ease the transition to Zope 4 and 5?

**Comment by datakurre** [comment]
@pbauer What do you think, if WebDAV without ZServer would be backported for Zope 4, would there be "final" Plone 5.x release for Python 2 with WebDAV on WSGI?

**Comment by dataflake** [comment]
I have no problem backporting it to Zope 4, no. Once it's done. There's still work to be done, like making the default object creation better (who needs DTML Documents anymore?), I need to look at the locking mechanism and its representation in the ZMI, and of course some real life testing.

File: src/ZPublisher/BaseRequest.py (hunk 1 above)
Comment (dataflake) [reply src/ZPublisher/BaseRequest.py]: It cannot due to circular imports. I tried.

**Comment by pbauer** [comment]
I could also imaging re-adding webdav-support to a 5.2.x release if its not a breaking change. Semantically 5.3 release would be correct though but I fear that would raise expectation to have more features that we're probably not able to live up to.

File: src/webdav/PropertySheets.py (hunk 2 above)
Comment (dataflake) [reply src/webdav/PropertySheets.py]: You're right, I deleted this one.

**Comment by dataflake** [comment]
The WebDAV reintegration PR is now ready for review. I will leave this open for a few days if anyone wants to suggest corrections or improvements.

If you go to https://github.com/zopefoundation/Zope/tree/webdav/src/webdav and scroll down you will also see testing notes from tests with the "litmus" WebDAV test tool, including tests against the first implementation in 2007. The current code is a bit better since I took the time to continue diagnosis on some of the issues @mcdonc found back then.

**Comment by pbauer** [comment]
I'd love to test this with Plone but we first have to make the branch for Plone 6 to run with the current state of Zope 5. We might find the time to do that at the Alpine City Sprint in Innsbruck next week. But please don't wait for us when it is ready merge!

**Comment by dataflake** [comment]
As soon as it's merged I will backport the whole thing to the `4.x` branch. 4.x needs a new release soon, anyway. I'll call that 4.2 then.

**Comment by dataflake** [comment]
FYI, Zope 4.2 is now released and it contains the WebDAV support code.

**Comment by zopyx** [comment]
How do you enable the WebDAV support in Zope 5.0a1? Setting webdav-source-port is enough or do you habe to make changes to the WSGI configuration?

**Comment by dataflake** [comment]
@zopyx It would be sufficient to ask this question in one place and not place the identical question in three places. Please read the documentation for `webdav-source-port` at https://zope.readthedocs.io/en/latest/operation.html#zope-configuration-reference

**Pull request merged**